mopidy/m3u/playlists.py

```diff
@@ -1,117 +1,117 @@
-        playlist = self._save_m3u(Playlist(name=name))
-        self._playlists[playlist.uri] = playlist
-        logger.info('Created playlist %s', playlist.uri)
-        return playlist
+        path = translator.path_from_name(name.strip(), self._default_extension)
+        try:
+            with self._open(path, 'w'):
+                pass
+            mtime = os.path.getmtime(self._abspath(path))
+        except EnvironmentError as e:
+            log_environment_error('Error creating playlist %s' % name, e)
+        else:
+            return translator.playlist(path, [], mtime)
 
     def delete(self, uri):
-        if uri in self._playlists:
-            path = translator.playlist_uri_to_path(uri, self._playlists_dir)
-            if os.path.exists(path):
-                os.remove(path)
-            else:
-                logger.warning(
-                    'Trying to delete missing playlist file %s', path)
-            del self._playlists[uri]
-            logger.info('Deleted playlist %s', uri)
+        path = translator.uri_to_path(uri)
+        try:
+            os.remove(self._abspath(path))
+        except EnvironmentError as e:
+            log_environment_error('Error deleting playlist %s' % uri, e)
+
+    def get_items(self, uri):
+        path = translator.uri_to_path(uri)
+        try:
+            with self._open(path, 'r') as fp:
+                items = translator.load_items(fp, self._playlists_dir)
+        except EnvironmentError as e:
+            log_environment_error('Error reading playlist %s' % uri, e)
         else:
-            logger.warning('Trying to delete unknown playlist %s', uri)
+            return items
 
     def lookup(self, uri):
-        return self._playlists.get(uri)
+        path = translator.uri_to_path(uri)
+        try:
+            with self._open(path, 'r') as fp:
+                items = translator.load_items(fp, self._playlists_dir)
+            mtime = os.path.getmtime(self._abspath(path))
+        except EnvironmentError as e:
+            log_environment_error('Error reading playlist %s' % uri, e)
+        else:
+            return translator.playlist(path, items, mtime)
 
     def refresh(self):
-        playlists = {}
-
-        encoding = sys.getfilesystemencoding()
-        for path in glob.glob(os.path.join(self._playlists_dir, b'*.m3u*')):
-            relpath = os.path.basename(path)
-            uri = translator.path_to_playlist_uri(relpath)
-            name = os.path.splitext(relpath)[0].decode(encoding, 'replace')
-            tracks = translator.parse_m3u(path)
-            playlists[uri] = Playlist(uri=uri, name=name, tracks=tracks)
+        pass  # nothing to do
 
-        self._playlists = playlists
-
-        logger.info(
-            'Loaded %d M3U playlists from %s',
-            len(playlists), self._playlists_dir)
+    def save(self, playlist):
+        path = translator.uri_to_path(playlist.uri)
+        name = translator.name_from_path(path)
+        try:
```

Comment:
We probably want to have a write-replace based on http://blog.gocept.com/2013/07/15/reliable-file-updates-with-python/ - My plan is currently to take the code in local, improve it and then expose in such a way that it can be used in local, for the state loading feature and here as well. Though I don't know yet if it would go into internal or not so maybe just copy the pattern from the link to have:

```
with tempfile.NamedTemporaryFile(
      'w', dir=os.path.dirname(filename), delete=False) as tf:
   tf.write(...)
   tempname = tf.name
os.rename(tempname, filename)
dirfd = os.open(os.path.dirname(filename), os.O_DIRECTORY)
os.fsync(dirfd)
os.close(dirfd)
```